### PR TITLE
Precisonat Bid Adapter : initial release

### PR DESCRIPTION
**Pull request by PrecisoSRL**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter  
- [ ] Updated bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

### CODE REVIEW

File: libraries/precisoUtils/bidUtils.js [hunk 1]
```diff
@@ -101,3 +88,156 @@ function macroReplace(adm, cpm) {
...
+    logInfo('Error in preciso pid fetch' + error);
+  }
+}
+
+function mapImpression(slot, bidderRequest) {
+  if (slot.mediaType === 'native' || deepAccess(slot, 'mediaTypes.native')) {
+    const imp = {
+      id: slot.bidId,
+      native: mapNative(slot),
+      bidfloor: slot.params.bidFloor,
```
Comment (bretg) [inline libraries/precisoUtils/bidUtils.js:110]: Sorry, not acceptable. You must first read the bid floor from the standard location as described in https://docs.prebid.org/dev-docs/bidder-adaptor.html . Specifically, [bidRequest.getFloor()](https://docs.prebid.org/dev-docs/modules/floors.html#bid-adapter-interface)

File: libraries/precisoUtils/bidUtils.js [hunk 2]
```diff
@@ -101,3 +88,156 @@ function macroReplace(adm, cpm) {
...
+      bidfloorcur: slot.params.currency
```
Comment (bretg) [inline libraries/precisoUtils/bidUtils.js:111]: Likewise, floor currency needs to come from the floor system so publishers don't have to specify it multiple times for different bidders

File: libraries/precisoUtils/bidUtils.js [hunk 3]
```diff
@@ -11,23 +12,7 @@ export const buildRequests = (endpoint) => (validBidRequests = [], bidderRequest
     // bidRequest: bidderRequest,
     id: validBidRequests[0].auctionId,
     cur: validBidRequests[0].params.currency || ['USD'],
```
Comment (bretg) [inline libraries/precisoUtils/bidUtils.js:14]: You need to first read currency out of config.getConfig(‘currency.adServerCurrency’). Your local params can be an alternate source.

File: modules/precisonatBidAdapter.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+Module that connects to preciso' demand sources
+please reach out to tech@preciso.net to receive your own publisher id
+
+# Parameters
+
+| Name          | Scope    | Description               | Example              |
+| :------------ | :------- | :------------------------ | :------------------- |
+| `publisherId` | required (for prebid.js) | partner ID provided by preciso | PreTest_0001
+| `region`   | required (for prebid.js)     | 3 letter country code | "USA" |
```
Comment (bretg) [inline modules/precisonatBidAdapter.md:19]: Region can't be required. It's difficult for global publishers to set this kind of metadata on their cached pages around the globe. You should be sniffing the IP address to get a sense for where in the world this user is.

**Review by bretg — CHANGES_REQUESTED** [review]

File: libraries/precisoUtils/bidUtils.js (hunk 1 above)
Comment (PrecisoSRL) [reply libraries/precisoUtils/bidUtils.js:110]: @bretg Thanks for the updates. I will change bid floor read by using [bidRequest.getFloor()](https://docs.prebid.org/dev-docs/modules/floors.html#bid-adapter-interface)

**Comment by PrecisoSRL** [comment]
Hi @bretg @ncolletti,

Thank you for the PR review and your valuable suggestions. To avoid code duplication, we have included native request and response handling in our existing bid adapter instead of creating a new one. I’ve implemented all the changes you suggested in the existing bid adapter as well. I’ve created another PR with those changes, and I will close this PR. 

you please take a look at the[ new PR](https://github.com/prebid/Prebid.js/pull/12257) when you have time?

Thank you!

File: modules/precisonatBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  isBidRequestValid: (bid) => {
+    logInfo('TESTETSTETSTETSTETST');
+    sharedId = storage.getDataFromLocalStorage('_sharedid') || storage.getCookie('_sharedid');
+    let precisoBid = true;
+    precisonatId = storage.getDataFromLocalStorage('_pre|id');
+    if (Object.is(precisonatId, 'NA') || Object.is(precisonatId, null) || Object.is(precisonatId, undefined)) {
+      if (!bid.precisoBid) {
+        precisoBid = false;
+        const puid = pid(sharedId);
```
[the remark this answers is not in the document]
Comment (bretg) [reply modules/precisonatBidAdapter.js:54]: @ncolletti - as long as the adapter is using the storagemanager to retrieve sharedId it seems ok to me.

**Comment by ChrisHuie** [comment]
closing with new pr

**Pull request closed without merging**